File: react/src/OnboardingView.js
import React, { useState, useEffect } from 'react';
import { View, Text, StyleSheet, TouchableOpacity, Image, Animated, Dimensions } from 'react-native';
import Swiper from 'react-native-swiper';
import Ionicons from 'react-native-vector-icons/Ionicons';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { requestLocationPermission, getCurrentLocation, isNearSanLuisObispo } from './OnboardingLocationManager';
import { useAdventureMode } from './AdventureModeContext';

const { width, height } = Dimensions.get('window');

const OnboardingView = ({ onComplete }) => {
  const [currentPage, setCurrentPage] = useState(0);
  const [hasAskedForLocation, setHasAskedForLocation] = useState(false);
  const [isAdventureModeRecommended, setIsAdventureModeRecommended] = useState(false);
  const [isAdventureModeEnabled, setIsAdventureModeEnabled] = useState(false);
  const { updateAdventureMode } = useAdventureMode();

  const adventureModeColor = '#4CAF50';
  const virtualTourColor = '#FF6803';

  const handleLocationPermission = async () => {
    const permissionGranted = await requestLocationPermission();
    setHasAskedForLocation(true);
    
    if (permissionGranted) {
      try {
        const position = await getCurrentLocation();
        const nearSLO = isNearSanLuisObispo(position);
        setIsAdventureModeRecommended(nearSLO);
        setIsAdventureModeEnabled(nearSLO);
      } catch (error) {
        console.error('Error getting location:', error);
        setIsAdventureModeRecommended(false);
        setIsAdventureModeEnabled(false);
      }
    } else {
      setIsAdventureModeRecommended(false);
      setIsAdventureModeEnabled(false);
    }
  };

  const handleComplete = async () => {
    await AsyncStorage.setItem('adventureMode', JSON.stringify(isAdventureModeEnabled));
    updateAdventureMode(isAdventureModeEnabled);
    onComplete(); // This will mark onboarding as complete and move to MainView
  };

  const renderWelcomeSlide = () => (
    <View style={styles.slide}>
      <Image source={require('../assets/icon.jpg')} style={styles.icon} />
      <View style={styles.titleContainer}>
        <Text style={styles.title}>Welcome to</Text>
        <Text style={[styles.title, styles.greenTitle, styles.boldTitle]}>Poly Canyon</Text>
      </View>
      <Text style={[styles.subtitle, styles.largerSubtitle]}>Explore and learn about Cal Poly's famous architectural structures</Text>
      <View style={styles.bottomButtonContainer}>
        {renderNavigationButton("Next", () => setCurrentPage(1))}
      </View>
    </View>
  );

  const renderLocationRequestSlide = () => (
    <View style={styles.slide}>
      <PulsingLocationDot />
      <View style={styles.titleContainer}>
        <Text style={styles.title}>Enable</Text>
        <Text style={[styles.title, styles.blueTitle, styles.boldTitle]}>Location Services</Text>
      </View>
      <Text style={[styles.subtitle, styles.largerSubtitle]}>We need your location to enhance your experience</Text>
      {!hasAskedForLocation ? (
        renderNavigationButton("Allow Location Access", handleLocationPermission)
      ) : (
        <View style={styles.bottomButtonContainer}>
          {renderNavigationButton("Next", () => setCurrentPage(2))}
        </View>
      )}
    </View>
  );

  const renderModeSelectionSlide = () => (
    <View style={styles.slide}>
      <Text style={[styles.title, styles.grayTitle, styles.largerTitle]}>Choose Your Experience</Text>
      <View style={styles.iconSpacing}>
        <ModeIcon
          name={isAdventureModeEnabled ? "walk" : "search"}
          color={isAdventureModeEnabled ? adventureModeColor : virtualTourColor}
        />
      </View>
      <CustomModePicker
        isAdventureModeEnabled={isAdventureModeEnabled}
        setIsAdventureModeEnabled={setIsAdventureModeEnabled}
        adventureModeColor={adventureModeColor}
        virtualTourColor={virtualTourColor}
      />
      <View style={styles.recommendationSpacing}>
        <RecommendationLabel isRecommended={isAdventureModeEnabled === isAdventureModeRecommended} />
      </View>
      <View style={styles.centeredFeatureList}>
        {isAdventureModeEnabled ? (
          <>
            <Text style={styles.largerFeatureItem}>• Explore structures in person</Text>
            <Text style={styles.largerFeatureItem}>• Track your progress</Text>
            <Text style={styles.largerFeatureItem}>• Use live location</Text>
          </>
        ) : (
          <>
            <Text style={styles.largerFeatureItem}>• Browse remotely</Text>
            <Text style={styles.largerFeatureItem}>• Learn about all structures</Text>
            <Text style={styles.largerFeatureItem}>• No location needed</Text>
          </>
        )}
      </View>
      <View style={styles.bottomButtonContainer}>
        <TouchableOpacity
          style={[styles.completeButton, { backgroundColor: isAdventureModeEnabled ? adventureModeColor : virtualTourColor }]}
          onPress={handleComplete}
        >
          <Text style={styles.completeButtonText}>Let's Go!</Text>
        </TouchableOpacity>
      </View>
    </View>
  );

  const ModeIcon = ({ name, color }) => (
    <View style={[styles.modeIcon, { backgroundColor: color }]}>
      <Ionicons name={name} size={40} color="white" />
    </View>
  );

  const CustomModePicker = ({ isAdventureModeEnabled, setIsAdventureModeEnabled, adventureModeColor, virtualTourColor }) => (
    <View style={styles.modePicker}>
      <TouchableOpacity
        style={[styles.modeButton, !isAdventureModeEnabled && styles.selectedMode]}
        onPress={() => setIsAdventureModeEnabled(false)}
      >
        <Text style={[styles.modeButtonText, !isAdventureModeEnabled && { color: virtualTourColor }]}>Virtual Tour</Text>
      </TouchableOpacity>
      <TouchableOpacity
        style={[styles.modeButton, isAdventureModeEnabled && styles.selectedMode]}
        onPress={() => setIsAdventureModeEnabled(true)}
      >
        <Text style={[styles.modeButtonText, isAdventureModeEnabled && { color: adventureModeColor }]}>Adventure</Text>
      </TouchableOpacity>
    </View>
  );

  const RecommendationLabel = ({ isRecommended }) => (
    <View style={[styles.recommendationLabel, { backgroundColor: isRecommended ? '#4CAF50' : '#FF5722' }]}>
      <Ionicons name={isRecommended ? "checkmark-circle" : "close-circle"} size={20} color="white" />
      <Text style={styles.recommendationText}>{isRecommended ? "Recommended" : "Not Recommended"}</Text>
    </View>
  );

  const PulsingLocationDot = () => {
    const pulseAnim = new Animated.Value(1);

    useEffect(() => {
      Animated.loop(
        Animated.sequence([
          Animated.timing(pulseAnim, { toValue: 1.2, duration: 1000, useNativeDriver: true }),
          Animated.timing(pulseAnim, { toValue: 1, duration: 1000, useNativeDriver: true })
        ])
      ).start();
    }, []);

    return (
      <View style={styles.pulsingDotContainer}>
        <Animated.View style={[styles.pulsingDot, { transform: [{ scale: pulseAnim }] }]} />
        <View style={styles.largerInnerDot} />
      </View>
    );
  };

  const renderNavigationButton = (text, onPress) => (
    <TouchableOpacity style={styles.navigationButton} onPress={onPress}>
      <Text style={styles.navigationButtonText}>{text}</Text>
      <Ionicons name="chevron-forward" size={24} color="white" />
    </TouchableOpacity>
  );

  return (
    <View style={styles.container}>
      <Swiper
        loop={false}
        showsPagination={true}
        index={currentPage}
        onIndexChanged={setCurrentPage}
        paginationStyle={styles.pagination}
        dotStyle={styles.dot}
        activeDotStyle={styles.activeDot}
      >
        {renderWelcomeSlide()}
        {renderLocationRequestSlide()}
        {renderModeSelectionSlide()}
      </Swiper>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'white',
  },
  slide: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 20,
  },
  icon: {
    width: 180,  // Increased size
    height: 180, // Increased size
    marginTop: 40,    // Less padding on top
    marginBottom: 20, // Less padding on bottom
    borderRadius: 36, // Adjusted for larger size
  },
  titleContainer: {
    marginBottom: 10,
  },
  title: {
    fontSize: 32,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  blackTitle: {
    color: 'black',
  },
  largerTitle: {
    fontSize: 36,
  },
  boldTitle: {
    fontWeight: '900',
  },
  greenTitle: {
    color: '#4CAF50',
  },
  blueTitle: {
    color: '#2196F3',
  },
  subtitle: {
    fontSize: 18,
    textAlign: 'center',
    marginBottom: 30,
    color: '#666',
  },
  largerSubtitle: {
    fontSize: 20,
  },
  modeIcon: {
    width: 80,
    height: 80,
    borderRadius: 40,
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 20,
  },
  modePicker: {
    flexDirection: 'row',
    backgroundColor: '#f0f0f0',
    borderRadius: 25,
    marginBottom: 20,
  },
  modeButton: {
    paddingVertical: 10,
    paddingHorizontal: 20,
    borderRadius: 25,
  },
  selectedMode: {
    backgroundColor: 'white',
  },
  modeButtonText: {
    fontSize: 16,
    fontWeight: 'bold',
  },
  recommendationLabel: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 5,
    paddingHorizontal: 10,
    borderRadius: 15,
    marginBottom: 20,
  },
  recommendationText: {
    color: 'white',
    marginLeft: 5,
    fontWeight: 'bold',
  },
  featureList: {
    alignSelf: 'stretch',
    marginBottom: 30,
  },
  featureItem: {
    fontSize: 16,
    marginBottom: 10,
  },
  completeButton: {
    paddingVertical: 12,
    paddingHorizontal: 30,
    borderRadius: 25,
  },
  completeButtonText: {
    color: 'white',
    fontSize: 18,
    fontWeight: 'bold',
  },
  pulsingDotContainer: {
    width: 100,
    height: 100,
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 30, // Added spacing below the circle
  },
  pulsingDot: {
    width: 100,
    height: 100,
    borderRadius: 50,
    backgroundColor: 'rgba(33, 150, 243, 0.3)',
    position: 'absolute',
  },
  largerInnerDot: {
    width: 70,
    height: 70,
    borderRadius: 35,
    backgroundColor: '#2196F3',
  },
  navigationButton: {
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: '#2196F3',
    paddingVertical: 10,
    paddingHorizontal: 20,
    borderRadius: 25,
    marginBottom: 30
  },
  navigationButtonText: {
    color: 'white',
    fontSize: 18,
    fontWeight: 'bold',
    marginRight: 10,
  },
  pagination: {
    bottom: 20,
  },
  dot: {
    width: 8,
    height: 8,
    borderRadius: 4,
    backgroundColor: '#bbb',
    margin: 3,
  },
  activeDot: {
    backgroundColor: '#2196F3',
  },
  bottomButtonContainer: {
    position: 'absolute',
    bottom: 60,
    left: 0,
    right: 0,
    alignItems: 'center', // Center horizontally
  },
  iconSpacing: {
    marginVertical: 20,
  },
  recommendationSpacing: {
    marginBottom: 10,
  },
  centeredFeatureList: {
    alignItems: 'center',
    marginBottom: 30,
  },
  largerFeatureItem: {
    fontSize: 18,
    marginBottom: 10,
  },
  grayTitle: {
    color: '#666', // Gray color
  },
});

export default OnboardingView;

File: swift/Poly Canyon/Views/OnboardingView.swift
import SwiftUI
import CoreLocation
import Shiny


struct OnboardingView: View {
    @Binding var isNewOnboardingCompleted: Bool
    @Binding var isAdventureModeEnabled: Bool
    @StateObject private var locationManager = OnboardingLocationManager()
    @State private var currentPage = 0
    @State private var hasAskedForLocation = false

    private let totalPages = 3
    let adventureModeColor = Color.green
    let virtualTourColor = Color(red: 255/255, green: 104/255, blue: 3/255, opacity: 1.0)

    var body: some View {
        TabView(selection: $currentPage) {
            welcomeSlide
                .tag(0)
            locationRequestSlide
                .tag(1)
            modeSelectionSlide
                .tag(2)
        }
        .tabViewStyle(PageTabViewStyle(indexDisplayMode: .never))
        .animation(.easeInOut, value: currentPage)
        .preferredColorScheme(.light)
    }

    var welcomeSlide: some View {
        VStack {
            Spacer()
            
                Image("Icon")
                    .resizable()
                    .scaledToFit()
                    .frame(width: 200, height: 200)
                    .cornerRadius(40)
                    .padding(.bottom, 30)
                    .shadow(color: .black.opacity(0.4), radius: 15)
                    .shiny()


            Text("Welcome to")
                .font(.system(size: 36))
                .foregroundColor(.black.opacity(1))
                .fontWeight(.bold)
            Text("Poly Canyon")
                .font(.system(size: 42))
                .foregroundColor(.green.opacity(1))
                .fontWeight(.bold)
                .padding(.bottom, 5)
            Text("Explore and learn about Cal Poly's famous architectural structures")
                .font(.system(size: 22))
                .foregroundColor(.black.opacity(0.8))
                .multilineTextAlignment(.center)
                .padding(.horizontal)
            Spacer()
            NavigationButton(text: "Next", action: {
                withAnimation {
                    currentPage = 1
                }
            })
            .padding(.bottom, 40)
        }
        .padding()
    }

    var locationRequestSlide: some View {
        VStack {
            Spacer()
            PulsingLocationDot()
                .frame(width: 100, height: 100)
                .padding(.bottom, 40)
            Text("Enable")
                .font(.system(size: 36))
                .foregroundColor(.black.opacity(1))
                .fontWeight(.bold)
            Text("Location Services")
                .font(.system(size: 38))
                .foregroundColor(.blue.opacity(1))
                .fontWeight(.bold)
                .padding(.bottom, 5)
            Text("We need your location to enhance your experience")
                .font(.system(size: 22))
                .foregroundColor(.black.opacity(0.8))
                .multilineTextAlignment(.center)
                .padding(.horizontal)
            Spacer()
            NavigationButton(text: locationManager.hasRequestedLocation ? "Next" : "Allow Location Access", action: {
                if !locationManager.hasRequestedLocation {
                    locationManager.requestLocation()
                } else {
                    withAnimation {
                        currentPage = 2
                    }
                }
            })
            .padding(.bottom, 40)
        }
        .padding()
    }

    var modeSelectionSlide: some View {
        VStack(spacing: 20) {
            Text("Choose Your Experience")
                .font(.system(size: 36, weight: .bold))
                .foregroundColor(.black.opacity(0.9))
                .multilineTextAlignment(.center)
                .padding(.top, 40)

            ModeIcon(
                imageName: isAdventureModeEnabled ? "figure.walk" : "binoculars",
                color: isAdventureModeEnabled ? adventureModeColor : virtualTourColor,
                isSelected: true
            )
            .padding(.vertical, 30)

            CustomModePicker(
                isAdventureModeEnabled: $isAdventureModeEnabled,
                adventureModeColor: adventureModeColor,
                virtualTourColor: virtualTourColor
            )
            .padding(.horizontal)

            RecommendationLabel(isRecommended: isAdventureModeEnabled == locationManager.isNearCalPoly)

            VStack(alignment: .leading, spacing: 10) {
                if isAdventureModeEnabled {
                    Text("• Explore structures in person")
                        .font(.system(size: 24, weight: .semibold))
                    Text("• Track your progress")
                        .font(.system(size: 24, weight: .semibold))
                    Text("• Use live location")
                        .font(.system(size: 24, weight: .semibold))
                } else {
                    Text("• Browse remotely")
                        .font(.system(size: 24, weight: .semibold))
                    Text("• Learn about all structures")
                        .font(.system(size: 24, weight: .semibold))
                    Text("• No location needed")
                        .font(.system(size: 24, weight: .semibold))
                }
            }
            .font(.system(size: 18))
            .foregroundColor(.black.opacity(0.8))
            .padding(.vertical)

            Spacer()

            Button("Let's Go!") {
                isNewOnboardingCompleted = true
            }
            .font(.system(size: 18, weight: .bold))
            .foregroundColor(.white)
            .frame(minWidth: 150)
            .padding(.vertical, 12)
            .padding(.horizontal, 30)
            .background(isAdventureModeEnabled ? adventureModeColor : virtualTourColor)
            .cornerRadius(25)
            .padding(.bottom, 40)
        }
        .padding()
        .background(Color.white)
        .onAppear {
            isAdventureModeEnabled = locationManager.isNearCalPoly
        }
    }
}
struct ModeIcon: View {
    let imageName: String
    let color: Color
    let isSelected: Bool

    var body: some View {
        ZStack {
            if isSelected {
                Circle()
                    .fill(color.opacity(0.3))
                    .frame(width: 100, height: 100)
                    .scaleEffect(1.05)
                    .animate()
            }
            
            Circle()
                .fill(color)
                .frame(width: 100, height: 100)



            Image(systemName: imageName)
                .resizable()
                .aspectRatio(contentMode: .fit)
                .frame(width: 60, height: 60)
                .foregroundColor(Color.white.opacity(0.8))
        }
    }
}

struct CustomModePicker: View {
    @Binding var isAdventureModeEnabled: Bool
    let adventureModeColor: Color
    let virtualTourColor: Color

    var body: some View {
        HStack(spacing: 0) {
            ModeButton(
                title: "Virtual Tour",
                isSelected: !isAdventureModeEnabled,
                color: virtualTourColor
            ) {
                withAnimation(.spring()) {
                    isAdventureModeEnabled = false
                }
            }

            ModeButton(
                title: "Adventure",
                isSelected: isAdventureModeEnabled,
                color: adventureModeColor
            ) {
                withAnimation(.spring()) {
                    isAdventureModeEnabled = true
                }
            }
        }
        .background(
            Capsule()
                .fill(Color.gray.opacity(0.2))
        )
        .overlay(
            Capsule()
                .stroke(Color.gray.opacity(0.3), lineWidth: 1)
        )
    }
}

struct ModeButton: View {
    let title: String
    let isSelected: Bool
    let color: Color
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(title)
                .font(.system(size: 20, weight: .bold))
                .padding(.vertical, 12)
                .padding(.horizontal, 20)
                .frame(maxWidth: .infinity)
                .background(
                    Capsule()
                        .fill(isSelected ? Color.white : Color.clear)
                        .shadow(color: isSelected ? Color.black.opacity(0.1) : Color.clear, radius: 5, x: 0, y: 2)
                )
                .foregroundColor(isSelected ? color : .gray)
        }
    }
}

struct RecommendationLabel: View {
    let isRecommended: Bool

    var body: some View {
        HStack {
            Image(systemName: isRecommended ? "checkmark.circle.fill" : "xmark.circle.fill")
                .foregroundColor(isRecommended ? .green : .red)
            Text(isRecommended ? "Recommended" : "Not Recommended")
                .font(.system(size: 16, weight: .semibold))
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 8)
        .background(Color.gray.opacity(0.1))
        .cornerRadius(20)
    }
}

struct PulsingLocationDot: View {
    @State private var scale: CGFloat = 1

    var body: some View {
        ZStack {
            Circle()
                .fill(Color.blue.opacity(0.25))
                .frame(width: 120, height: 120)
                .scaleEffect(scale)
                .animation(
                    Animation.easeInOut(duration: 1)
                        .repeatForever(autoreverses: true),
                    value: scale
                )
            Circle()
                .fill(Color.blue)
                .frame(width: 100, height: 100)
        }
        .onAppear {
            self.scale = 1.2
        }
    }
}

struct NavigationButton: View {
    let text: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack {
                Text(text)
                    .font(.system(size: 18, weight: .bold))
                Image(systemName: "chevron.right")
                    .font(.system(size: 14, weight: .bold))
            }
            .padding(.horizontal, 20)
            .padding(.vertical, 12)
            .background(Color.black.opacity(0.8))
            .foregroundColor(.white)
            .cornerRadius(25)
        }
    }
}

extension View {
    func animate() -> some View {
        self.modifier(PulseAnimation())
    }
}

struct PulseAnimation: ViewModifier {
    @State private var animationAmount: CGFloat = 1

    func body(content: Content) -> some View {
        content
            .scaleEffect(animationAmount)
            .animation(
                Animation.easeInOut(duration: 1.5)
                    .repeatForever(autoreverses: true),
                value: animationAmount
            )
            .onAppear {
                self.animationAmount = 1.2
            }
    }
}

struct OnboardingView_Previews: PreviewProvider {
    static var previews: some View {
        OnboardingView(
            isNewOnboardingCompleted: .constant(false),
            isAdventureModeEnabled: .constant(false)
        )
    }
}
